File: cpp/456.132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int lastMin = std::numeric_limits<int>::max();
        stack<pair<int, int>> st;
        
        for (auto num : nums) {

            while (!st.empty() && st.top().first < num) {
                st.pop();
            }

            if (!st.empty() && st.top().first > num && num > st.top().second) {
                return true;
            }

            st.push(make_pair(num, lastMin));
            lastMin = min(lastMin, num);
        }
        return false;
    }
};

File: python/456.132-pattern.py
from bisect import bisect


class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        # n = len(nums)
        # for i in range(n):
        #     for j in range(n):
        #         for k in range(n):
        #             if i < j < k and  nums[i] < nums[k] < nums[j]:
        #                 print(i,j, k)
        #                 print(nums[i], nums[k], nums[j])
        #                 return True
        st = []
        last = float("inf")
        for num in nums:

            while st and st[-1][0] < num:
                st.pop()

            if st and st[-1][0] > num and num > st[-1][1]:
                return True

            st.append((num, last))
            last = min(last, num)

        return False